File: kordbooks_erp/kordbooks_practice/page/sales_invoice_entry/sales_invoice_entry.js
frappe.pages["sales-invoice-entry"].on_page_load = function (wrapper) {
	let page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Sales Invoice Entry",
		single_column: true,
	});

	let $main = $(wrapper).find(".layout-main-section");

	$main.html(`
		<div class="kordbooks-sales-entry" style="padding: 16px; max-width: 760px;">
			<h3 style="margin-bottom: 16px;">Sales Invoice Entry</h3>

			<div class="row" style="margin-bottom: 12px;">
				<div class="col-sm-6">
					<label class="control-label">Company</label>
					<input type="text" id="company" class="form-control" placeholder="Company name">
				</div>
				<div class="col-sm-6">
					<label class="control-label">Invoice Number</label>
					<input type="text" id="invoice_number" class="form-control" placeholder="External invoice number">
				</div>
			</div>

			<div class="row" style="margin-bottom: 12px;">
				<div class="col-sm-4">
					<label class="control-label">Invoice Date</label>
					<input type="date" id="invoice_date" class="form-control">
				</div>
				<div class="col-sm-4">
					<label class="control-label">Payment Date</label>
					<input type="date" id="payment_date" class="form-control">
				</div>
				<div class="col-sm-4">
					<label class="control-label">Amount</label>
					<input type="number" step="0.01" id="amount" class="form-control" placeholder="0.00">
				</div>
			</div>

			<div class="row" style="margin-bottom: 12px;">
				<div class="col-sm-6">
					<label class="control-label">Shortcut</label>
					<select id="shortcut" class="form-control">
						<option value="food">Food Sales</option>
						<option value="drinks">Drink Sales</option>
						<option value="accommodation">Accommodation</option>
						<option value="service_charge">Service Charge</option>
						<option value="general">General Sales</option>
					</select>
				</div>
				<div class="col-sm-6">
					<label class="control-label">VAT Mode</label>
					<select id="vat_mode" class="form-control">
						<option value="no_vat">No VAT</option>
						<option value="vat_20">VAT 20%</option>
					</select>
				</div>
			</div>

			<div style="margin-bottom: 16px;">
				<label class="control-label">Notes</label>
				<textarea id="notes" class="form-control" rows="3" placeholder="Optional notes"></textarea>
			</div>

			<div style="display: flex; gap: 8px;">
				<button class="btn btn-primary" id="save-draft">Save Draft</button>
				<button class="btn btn-default" id="save-submit">Save & Submit</button>
			</div>
		</div>
	`);

	function collect_data() {
		return {
			company: $main.find("#company").val(),
			invoice_number: $main.find("#invoice_number").val(),
			invoice_date: $main.find("#invoice_date").val(),
			payment_date: $main.find("#payment_date").val(),
			amount: $main.find("#amount").val(),
			shortcut: $main.find("#shortcut").val(),
			vat_mode: $main.find("#vat_mode").val(),
			notes: $main.find("#notes").val(),
		};
	}

	function save_invoice(submit_after_insert) {
		frappe.call({
			method: "kordbooks_erp.kordbooks_practice.page.sales_invoice_entry.sales_invoice_entry.create_sales_invoice",
			args: {
				data: collect_data(),
				submit_after_insert: submit_after_insert ? 1 : 0,
			},
			freeze: true,
			freeze_message: "Creating Sales Invoice...",
			callback: function (r) {
				if (!r.message) return;

				frappe.msgprint(`Created Sales Invoice: ${r.message.name}`);

				if (r.message.name) {
					frappe.set_route("Form", "Sales Invoice", r.message.name);
				}
			},
		});
	}

	$main.find("#save-draft").on("click", function () {
		save_invoice(false);
	});

	$main.find("#save-submit").on("click", function () {
		save_invoice(true);
	});
};

File: kordbooks_erp/kordbooks_practice/page/sales_invoice_entry/sales_invoice_entry.py
# This code is not a new DocType. It is backend Python code for your custom page, and its job is to create a standard ERPNext Sales Invoice doctype by using Frappe’s Document API

import frappe
from frappe.utils import flt, getdate


DEFAULT_CUSTOMER = "Retail Sales Customer"

SHORTCUTS = {
	"food": {
		"item_code": "FOOD-SALES",
		"description": "Food Sales",
	},
	"drinks": {
		"item_code": "DRINKS-SALES",
		"description": "Drink Sales",
	},
	"accommodation": {
		"item_code": "ROOM-SALES",
		"description": "Accommodation Sales",
	},
	"service_charge": {
		"item_code": "SERVICE-CHARGE",
		"description": "Service Charge",
	},
	"general": {
		"item_code": "GENERAL-SALES",
		"description": "General Sales",
	},
}

VAT_TEMPLATES = {
	"no_vat": None,
	"vat_20": "UK VAT 20%",
}


def _validate_payload(data):
	if not data.get("company"):
		frappe.throw("Company is required")

	if not data.get("invoice_date"):
		frappe.throw("Invoice Date is required")

	if flt(data.get("amount")) <= 0:
		frappe.throw("Amount must be greater than zero")

	if data.get("shortcut") not in SHORTCUTS:
		frappe.throw("Invalid shortcut selected")

	if not frappe.db.exists("Customer", DEFAULT_CUSTOMER):
		frappe.throw(
			f"Default customer '{DEFAULT_CUSTOMER}' does not exist. Create it first."
		)

	item_code = SHORTCUTS[data.get("shortcut")]["item_code"]
	if not frappe.db.exists("Item", item_code):
		frappe.throw(f"Item '{item_code}' does not exist. Create it first.")


@frappe.whitelist()
def create_sales_invoice(data, submit_after_insert=0):
	data = frappe.parse_json(data)
	_validate_payload(data)

	shortcut = SHORTCUTS[data.get("shortcut")]
	tax_template = VAT_TEMPLATES.get(data.get("vat_mode"))

	doc = frappe.get_doc(
		{
			"doctype": "Sales Invoice",
			"company": data.get("company"),
			"customer": DEFAULT_CUSTOMER,
			"posting_date": getdate(data.get("invoice_date")),
			"due_date": getdate(data.get("payment_date") or data.get("invoice_date")),
			"taxes_and_charges": tax_template,
			"kordbooks_invoice_number": data.get("invoice_number"),
			"kordbooks_payment_date": getdate(data.get("payment_date"))
			if data.get("payment_date")
			else None,
			"kordbooks_shortcut": data.get("shortcut"),
			"kordbooks_notes": data.get("notes"),
			"items": [
				{
					"item_code": shortcut["item_code"],
					"description": shortcut["description"],
					"qty": 1,
					"rate": flt(data.get("amount")),
				}
			],
		}
	)

	doc.insert()

	if int(submit_after_insert):
		doc.submit()

	return {
		"name": doc.name,
		"docstatus": doc.docstatus,
	}